### DISCUSSION TITLE
Провайдер на слое app

### DISCUSSION THREAD

**Discussion by khanov26** [feature-sliced/documentation#694] [Q&A] [answered]
Здравствуйте!

Есть провайдер на слое app.
<img width="356" alt="Снимок экрана 2024-07-17 в 1 23 21 PM" src="https://github.com/user-attachments/assets/3554b638-d1ea-4b64-9a6c-7a7883125137">

`model/context.ts`
```
import { createContext, useContext } from 'react'
import { CurrenciesContextType } from './types'

const defaultValue: CurrenciesContextType = {
    isFetching: false,
    currencies: {},
    error: ''
}

export const CurrenciesContext =
    createContext<CurrenciesContextType>(defaultValue)

export function useCurrencies() {
    return useContext(CurrenciesContext)
}
```

`ui/Provider.tsx`
```
import { FC, ReactNode } from 'react'
import { useFetchCurrencies } from '../api/useFetchCurrencies'
import { CurrenciesContext } from '../model/context'

interface Props {
    children: ReactNode
}

export const CurrenciesProvider: FC<Props> = ({ children }) => {
    const { data, isFetching, error } = useFetchCurrencies()

    return (
        <CurrenciesContext.Provider
            value={{ currencies: data, isFetching, error }}
        >
            {children}
        </CurrenciesContext.Provider>
    )
}
```

Однако, чтобы получить данные на слоях ниже, мне приходится импортировать useCurrencies из верхнего слоя app, что противоречит FSD.
Можно CurrenciesContext и useCurrencies перенести на слой share, но тогда это приведет к дроблению модуля, и не будет никакой связности кода.

**Comment by illright** [comment]
Здравствуйте, немного не понял, в чем заключается вопрос. Обычно, если провайдер лежит в app, то он либо предоставляет библиотечную функциональность (стили или сторы), либо он провайдит контекст, который лежит на слое ниже (чаще всего, Shared). Вопрос хранения связанных модулей рядом актуальный, но его решение зависит от изначальной цели этого контекста. Нужно разобраться в кейсе, поэтому возможно, лучше задать этот вопрос в Телеграме, где люди могут спросить уточняющие вопросы

**Reply by khanov26** [reply]
Т.е. вы предлагаете создать контекст _CurrenciesContext_ в share, а в _CurrenciesProvider_ импортировать этот контекст? Я не понимаю, зачем разделять тесно связанные между собой провайдер и контекст на разные слои. Будет ли ошибкой, если я перенесу провайдер в слой share?

Цель провайдера _CurrenciesProvider_ - получить данные по сети один раз при запуске приложения и провайдить эти данные на все приложение.

**Reply by illright** [reply, marked as answer]
Нет, ошибкой не будет
